frontend/packages/client/src/components/StepByStep/index.js

```diff
@@ -32,25 +38,80 @@ function StepByStep({
 
   const [currentStep, setCurrentStep] = useState(0);
   const [showPreStep, setShowPreStep] = useState(!!preStep);
-  const [isStepValid, setStepValid] = useState(false);
+  const [isStepValid, _setStepValid] = useState(false);
   const [stepsData, setStepsData] = useState({});
-  const refs = React.useRef();
+  const refs = useRef();
+  const [validSteps, setValidSteps] = useState([]);
+  const setStepValid = (validity) => {
+    if (validity) {
+      if (!validSteps.length) {
+        setValidSteps([currentStep]);
+      } else if (!validSteps.includes(currentStep)) {
+        setValidSteps([...validSteps, currentStep]); // add validated
+      } // else already included, so do nothing
+    } else if (validSteps.includes(currentStep)) {
+      setValidSteps(validSteps.filter(step => step === currentStep)); // remove invalidated
+    } // else invalid and not included anyways, so do nothing
+    _setStepValid(validity);
+  }
+  const [searchParams] = useSearchParams();
+  const [SPCopy, setSPC] = useState("1");
+  const { pathname, search } = useLocation();
+  const navigate = useNavigate();
+  const [once, setOnce] = useState();
 
-  const onStepAdvance = (direction = "next") => {
-    if (direction === "next") {
-      if (currentStep + 1 <= steps.length - 1) {
-        const enableAdvance = runPreCheckStepAdvance();
-        if (!enableAdvance) {
-          return;
-        }
-        setCurrentStep(currentStep + 1);
-      }
-    } else if (direction === "prev") {
-      if (currentStep - 1 >= 0) {
-        setCurrentStep(currentStep - 1);
+  useEffect(() => {
+    const displayStep = searchParams.get("step");
+    if (once && (displayStep !== SPCopy)) { // navigated to a new step
+      setSPC(displayStep);
+      setOnce(false);
+    }
+  }, [SPCopy, once, searchParams]);
+
+  const goToStep = useCallback((_step, replace = false) => {
+    const step = _step + 1;
+    let newQuery;
+    const displayStepLoc = search.indexOf("step=");
+    if (displayStepLoc === -1) { // no display step yet
+      newQuery = search.concat(`&step=${step}`);
+    } else {
+      newQuery = search.slice(0, displayStepLoc + "step=".length).concat(step);
+    }
```

Comment:
This could be simplified with the usage of `URLSearchParams` ie
```js
const q = new URLSearchParams(search)
q.set('step', x)
return q.toString()
```